src/session/client-certificate.c

```diff
@@ -127,6 +132,97 @@ unsigned long long get_proc_pid_start_time (int dirfd)
   return start_time;
 }
 
+/* Fallback for get_ws_proc_fd() on older kernels which don't support enough pidfd API */
+static int
+get_ws_proc_fd_pid_time (void)
+{
+  struct ucred ucred;
+  socklen_t ucred_len = sizeof ucred;
+  if (getsockopt (STDIN_FILENO, SOL_SOCKET, SO_PEERCRED, &ucred, &ucred_len) != 0 ||
+      /* this is an inout parameter, be extra suspicious */
+      ucred_len < sizeof ucred)
+    {
+      debug ("failed to read stdin peer credentials: %m; not in socket mode?");
+      warnx ("Certificate authentication only supported with cockpit-session.socket");
+      exit_init_problem ("authentication-unavailable", "Certificate authentication only supported with cockpit-session.socket");
+    }
+
+  debug ("unix socket mode, ws peer pid %d", ucred.pid);
+  int ws_pid_dirfd = open_proc_pid (ucred.pid);
+  unsigned long long ws_start_time = get_proc_pid_start_time (ws_pid_dirfd);
+
+  int my_pid_dirfd = open_proc_pid (getpid ());
+  unsigned long long my_start_time = get_proc_pid_start_time (my_pid_dirfd);
+  close (my_pid_dirfd);
+
+  debug ("peer start time: %llu, my start time: %llu", ws_start_time, my_start_time);
+
+  /* Guard against pid recycling: If a malicious user captures ws, keeps the socket in a forked child and exits
+    * the original pid, they can trap a different user to login, get the old pid (pointing ot their cgroup), and
+    * capture their session. To prevent that, require that ws must have started earlier than ourselves. */
+  if (my_start_time < ws_start_time)
+    {
+      warnx ("start time of this process (%llu) is older than cockpit-ws (%llu), pid recycling attack?",
+              my_start_time, ws_start_time);
+      close (ws_pid_dirfd);
+      exit_init_problem ("access-denied", "implausible cockpit-ws start time");
+    }
+
+  return ws_pid_dirfd;
+}
+
+/* Get a /proc/[pid] dirfd for our Unix socket peer (i.e. cockpit-ws).
+ * We only support being called via cockpit-session.socket (i.e. Unix socket).
+ */
+static int
+get_ws_proc_fd (void)
+{
+#if defined(SO_PEERPIDFD) && defined(HAVE_PIDFD_GETPID)
+  int pidfd = -1;
+  socklen_t socklen = sizeof pidfd;
+  /* this is always the pidfd for the process that started the communication, it cannot be recycled */
+  if (getsockopt (STDIN_FILENO, SOL_SOCKET, SO_PEERPIDFD, &pidfd, &socklen) < 0)
+    {
+      if (errno == ENOPROTOOPT || errno == ENODATA)
```

Comment:
The `ENODATA` fallback here feels inappropriate.  We want only to detect the case where the kernel doesn't support this feature yet (ie: `ENOPROTOOPT`).  The case where the specific socket type doesn't support this query is highly unexpected and should cause a hard bomb.

Is there a reason you included this one?

See https://lore.kernel.org/all/CAMw=ZnTcsYXnwDcKNxwbLCE3dWwaw_UBRmaBUNCYEF5X4BCGuw@mail.gmail.com/ about `ENODATA`.

Reply:
Mostly just because https://github.com/polkit-org/polkit/pull/501 used it as "valid" error code. But I'm happy to lose it, CI will tell us if we are wrong.

Reply:
Killed ENODATA. I don't expect CI failures, but if there are any, I'll investigate.

Reply:
On the principle of "be strict until we find out it's a problem" could we drop the entire fallback code path here (at the `#ifdef` level, even) if we have `SO_PEERPIDFD` and `pidfd_getpid()`?

Put another way: is this runtime `ENOPROTOOPT` case hit on any of our supported OSes?

Reply:
> On the principle of "be strict until we find out it's a problem" could we drop the entire fallback code path here (at the `#ifdef` level, even) if we have `SO_PEERPIDFD` and `pidfd_getpid()`?

That's a build time vs runtime check. Even if the program was built against a set of headers that have those APIs, it doesn't necessarily mean it will run on a kernel that has them

Reply:
I understand the theoretical possibility but my question is about our supported OSes...

I guess you're looking out for people running "other" configurations, of course... just seems weird to have the fallback *only* for people who we don't officially support...

Reply:
Well, we also call the fallback for "not present at build time", but indeed this particular error path we don't cover in tests. Nevertheless, I think we should keep it.

Reply:
nix dagegen